pyopencl/invoker.py

```diff
@@ -305,6 +306,65 @@ def gen_arg_setting(in_enqueue):
             enqueue_name)
 
 
+# {{{ Helper functions related to argument sizes and device limits
+
+def _get_max_parameter_size(dev):
+    """Return the device's maximum parameter size adjusted for pocl."""
+    from pyopencl.characterize import get_pocl_version
+
+    dev_limit = dev.max_parameter_size
+
+    if get_pocl_version(dev.platform) is not None:
+        # Older pocl versions have an incorrect parameter size limit of 1024
+        # See e.g. https://github.com/pocl/pocl/pull/1046
+        if dev_limit == 1024:
+            if dev.type & cl.device_type.CPU:
+                return 1024*1024
+            if dev.type & cl.device_type.GPU:
+                # All modern Nvidia GPUs (starting from Compute Capability 2)
+                # have this limit
+                return 4352
+
+    return dev_limit
+
+
+def _check_arg_size(function_name, num_cl_args, arg_types, devs):
+    """Check whether argument sizes exceed the OpenCL device limit."""
+
+    for dev in devs:
+        dev_ptr_size = int(dev.address_bits / 8)
+        dev_limit = _get_max_parameter_size(dev)
+
+        total_arg_size = 0
+
+        if arg_types:
+            for arg_type in arg_types:
+                if arg_type is None or isinstance(arg_type, VectorArg):
+                    total_arg_size += dev_ptr_size
+                else:
+                    total_arg_size += np.dtype(arg_type).itemsize
+        else:
+            # Estimate that each argument has the size of a pointer on average
+            total_arg_size = dev_ptr_size * num_cl_args
+
+        if total_arg_size > dev_limit:
+            from warnings import warn
+            warn(f"Kernel '{function_name}' has {num_cl_args} arguments with "
+                f"a total size of {total_arg_size} bytes, which is higher than "
+                f"the limit of {dev_limit} bytes on {dev}. This might "
+                "lead to compilation errors especially on CUDA devices.")
+        elif total_arg_size >= dev_limit * 0.75:
+            # Since total_arg_size is just an estimate, also warn in case we are
```

Comment:
In some cases it's exact, in others it's an estimate. This should only trigger if we're estimating.

Reply:
What do you think of [db9bef2](https://github.com/inducer/pyopencl/pull/555/commits/db9bef2b90e08a7fc64ddab68c93aeeb69570273)?

Reply:
Yep, that's what I had in mind.